test/source/allocation.cpp

```diff
@@ -19,13 +19,79 @@ auto get_one(smeagle::Corpus const& corpus, char const* name) {
   return funcs;
 }
 
-TEST_CASE("Register Allocation") {
+TEST_CASE("Register Allocation and Types") {
   auto corpus = smeagle::Smeagle("liballocation.so").parse();
 
   SUBCASE("one int") {
-    auto funcs = get_one(corpus, "f1");
+    auto funcs = get_one(corpus, "test_type_int");
     REQUIRE(funcs.size() == 1);
     auto const& parameters = funcs[0].parameters;
+    CHECK(parameters[0].type == "int");
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("one string") {
+    auto funcs = get_one(corpus, "test_type_string");
+    REQUIRE(funcs.size() == 1);
+    auto const& parameters = funcs[0].parameters;
+    CHECK(parameters[0].type == "string");
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("one char") {
+    auto funcs = get_one(corpus, "test_type_char");
+    REQUIRE(funcs.size() == 1);
+    auto const& parameters = funcs[0].parameters;
+    CHECK(parameters[0].type == "char");
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("one bool") {
+    auto funcs = get_one(corpus, "test_type_boolean");
+    REQUIRE(funcs.size() == 1);
+    auto const& parameters = funcs[0].parameters;
+    CHECK(parameters[0].type == "bool");
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("void") {
+    auto funcs = get_one(corpus, "test_type_void");
+    REQUIRE(funcs.size() == 1);
+    auto const& parameters = funcs[0].parameters;
+    REQUIRE(parameters.size() == 0);
+  }
+
+  SUBCASE("one wchar_t") {
```

Comment:
This case segfaults. It actually brings up a good point: what about Unicode? The AMD64 ABI doc says nothing about it. The C++ standard says only that `wchar_t` should "be large enough to represent any supported character code point". This could be UTF-16 or UTF-32 (aka Unicode). The changes I'm adding to Dyninst will provide support for detecting UTF encodings in DWARF4 or newer. However, I'm not sure what the answer should be for how a register should be allocated. My _guess_ would be the same as a 32-bit int.